src/uct/tcp/tcp_listener.c

```diff
@@ -0,0 +1,216 @@
+    int backlog;
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_listener_t, cm);
+
+    self->conn_request_cb = params->conn_request_cb;
+    self->user_data       = (params->field_mask & UCT_LISTENER_PARAM_FIELD_USER_DATA) ?
+                            params->user_data : NULL;
+    backlog               = (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) ?
+                            params->backlog : SOMAXCONN;
+
+    status = ucs_socket_create(saddr->sa_family, SOCK_STREAM, &self->listener_fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the listener_fd to non-blocking mode
+     * (so that accept() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(self->listener_fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    if (bind(self->listener_fd, saddr, socklen) < 0) {
+        ucs_error("bind(fd=%d addr=%s) failed: %m",
+                  self->listener_fd, ucs_sockaddr_str(saddr, ip_port_str,
+                                                      UCS_SOCKADDR_STRING_LEN));
+        status = (errno == EADDRINUSE) ? UCS_ERR_BUSY : UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    if (listen(self->listener_fd, backlog) < 0) {
+        ucs_error("listen(fd=%d addr=%s backlog=%d) failed: %m",
+                  self->listener_fd, ucs_sockaddr_str(saddr, ip_port_str,
+                                                      UCS_SOCKADDR_STRING_LEN),
+                  backlog);
+        status = UCS_ERR_IO_ERROR;
+        goto err_close_socket;
+    }
+
+    ucs_list_head_init(&self->fd_list);
+
+    async_ctx = tcp_sockcm->super.iface.worker->async;
+    status = ucs_async_set_event_handler(async_ctx->mode, self->listener_fd,
+                                         UCS_EVENT_SET_EVREAD |
+                                         UCS_EVENT_SET_EVERR,
+                                         uct_tcp_listener_conn_req_handler, self,
+                                         async_ctx);
+    if (status != UCS_OK) {
+        goto err_close_socket;
+    }
+
+    ucs_debug("created a TCP listener %p on cm %p with fd: %d. "
+              "listening on %s", self, cm, self->listener_fd,
+              ucs_sockaddr_str(saddr, ip_port_str, UCS_SOCKADDR_STRING_LEN));
+
+    return UCS_OK;
+
+err_close_socket:
+    close(self->listener_fd);
+err:
+    return status;
+}
+
+UCS_CLASS_CLEANUP_FUNC(uct_tcp_listener_t)
+{
+    uct_tcp_listener_ctx_t *listen_ctx, *tmp;
+    ucs_status_t status;
+
+    status = ucs_async_remove_handler(self->listener_fd, 1);
+    if (status != UCS_OK) {
+        ucs_warn("failed to remove event handler for fd %d: %s",
+                 self->listener_fd, ucs_status_string(status));
+    }
+
+    close(self->listener_fd);
+
+    UCS_ASYNC_BLOCK(self->super.cm->iface.worker->async);
+
+    ucs_list_for_each_safe(listen_ctx, tmp, &self->fd_list, list) {
+        ucs_async_remove_handler(listen_ctx->fd, 1);
```

Comment:
shouldn't we handle it or warn about this?

Reply:
handle what? this is the listener cleanup

Reply:
is there clean up of not handled events?

Reply:
which events for example? handled how?

Reply:
seems line I don't completely understand the meaning of this list... It grows per every connect request and cleaned up only on listener destroy?

Reply:
currently yes. will probably need to remove an fd on disconnect events too but this will be done later when eps are implemented.

Reply:
then should not this fd be part of EP and no need to put it on the listener list

Reply:
yes but tcp_sockcm_ep isn't implemented in this pr but the handler for connection request from the client is created here since it's on the listener. so if i add an fd to the handler, i have to remove it too.

Reply:
I think you have to pass it to user callback as part of connection request and put it on EP in next PR with EP implementation. the same way as we do in rdmacm_cm with rdmacm_id

Reply:
BTW `accept` also should be called from ep_create on server side. Otherwise you have fd which can generate EP related events but EP is not created yet.

Reply:
accept can't be called before an ep exists but i will create an ep in this function, before ep_create is called, and it will become valid once uct_ep_create is called (will be passed in ep params). 
but this pr doesn't have eps...

Reply:
you are going to pass ep as part of ep_params? something wrong here.

Reply:
it will be part of conn_request along with the fd and then passed to ep_create. why is this a problem?

Reply:
how reject will be look like? user should deicide accept or reject incoming connection (excepting error handling flow)

Reply:
what's the problem with reject?
i think we can move this discussion to the ep/reject pr when it's ready...

Reply:
Ok, my concern regarding to listener is the list of fds looks redundant on listener.